### PR TITLE
[ticket/11444] Moving the in-board notifications to a method class

### PR DESCRIPTION
**Pull request by Nicofuma**
Currently the in-board method for the notifications is hardcoded and
cannot be disabled. This method should be in his own class extending
`phpbb\notification\method\method_interface`.

It also add the possibility, for each method, to be enabled by default (ie:
no entry in the DB => notification enabled).

https://tracker.phpbb.com/browse/PHPBB3-11444
https://tracker.phpbb.com/browse/PHPBB3-11967

Replaces https://github.com/phpbb/phpbb/pull/2365

PHPBB3-11967
PHPBB3-11444

### CODE REVIEW

File: phpBB/phpbb/notification/manager.php [hunk 1]
```diff
@@ -13,7 +13,7 @@
 
 namespace phpbb\notification;
 
-use Symfony\Component\DependencyInjection\ContainerInterface;
+use \Symfony\Component\DependencyInjection\ContainerInterface;
```
Comment (CHItA) [inline phpBB/phpbb/notification/manager.php]: ?

File: phpBB/phpbb/notification/method/board.php [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+	public function __construct(\phpbb\user_loader $user_loader, \phpbb\db\driver\driver_interface $db, \phpbb\cache\driver\driver_interface $cache, $user, \phpbb\auth\auth $auth, \phpbb\config\config $config, $phpbb_root_path, $php_ext, $notification_types_table, $notifications_table)
+	{
+		$this->user_loader = $user_loader;
+		$this->db = $db;
+		$this->cache = $cache;
+		$this->user = $user;
+		$this->auth = $auth;
+		$this->config = $config;
+		$this->phpbb_root_path = $phpbb_root_path;
+		$this->php_ext = $php_ext;
```
Comment (CHItA) [inline phpBB/phpbb/notification/method/board.php]: auth, phpbb_root_path and php_ext seems not to be used. Maybe inject it to base with a setter?

Comment (CHItA) [inline phpBB/phpbb/notification/method/board.php]: Actually I would prefer injecting all base dependencies to the base, and using inheritance in the config files

File: phpBB/phpbb/notification/type/base.php [hunk 3]
```diff
@@ -100,11 +97,10 @@
 	* @param string $phpbb_root_path
 	* @param string $php_ext
 	* @param string $notification_types_table
-	* @param string $notifications_table
 	* @param string $user_notifications_table
 	* @return \phpbb\notification\type\base
 	*/
-	public function __construct(\phpbb\user_loader $user_loader, \phpbb\db\driver\driver_interface $db, \phpbb\cache\driver\driver_interface $cache, $user, \phpbb\auth\auth $auth, \phpbb\config\config $config, $phpbb_root_path, $php_ext, $notification_types_table, $notifications_table, $user_notifications_table)
+	public function __construct(\phpbb\user_loader $user_loader, \phpbb\db\driver\driver_interface $db, \phpbb\cache\driver\driver_interface $cache, $user, \phpbb\auth\auth $auth, \phpbb\config\config $config, $phpbb_root_path, $php_ext, $notification_types_table, $user_notifications_table)
```
Comment (CHItA) [inline phpBB/phpbb/notification/type/base.php]: Is it necessary to have the unused dependencies here? Also applies to the `method\base`

File: phpBB/phpbb/notification/manager.php [hunk 4]
```diff
@@ -80,12 +77,11 @@ class manager
 	* @param string $phpbb_root_path
 	* @param string $php_ext
 	* @param string $notification_types_table
-	* @param string $notifications_table
 	* @param string $user_notifications_table
 	*
 	* @return \phpbb\notification\manager
 	*/
-	public function __construct($notification_types, $notification_methods, ContainerInterface $phpbb_container, \phpbb\user_loader $user_loader, \phpbb\config\config $config, \phpbb\event\dispatcher_interface $phpbb_dispatcher, \phpbb\db\driver\driver_interface $db, \phpbb\cache\service $cache, $user, $phpbb_root_path, $php_ext, $notification_types_table, $notifications_table, $user_notifications_table)
+	public function __construct($notification_types, $notification_methods, ContainerInterface $phpbb_container, \phpbb\user_loader $user_loader, \phpbb\config\config $config, \phpbb\event\dispatcher_interface $phpbb_dispatcher, \phpbb\db\driver\driver_interface $db, \phpbb\cache\service $cache, $user, $phpbb_root_path, $php_ext, $notification_types_table, $user_notifications_table)
```
Comment (CHItA) [inline phpBB/phpbb/notification/manager.php]: This class also has unused dependencies.

File: phpBB/phpbb/notification/method/base.php [hunk 5]
```diff
@@ -104,6 +134,55 @@ public function add_to_queue(\phpbb\notification\type\type_interface $notificati
 	}
 
 	/**
+	* {@inheritdoc}
+	*/
+	public function update_notification($notification, array $data, array $options)
```
Comment (CHItA) [inline phpBB/phpbb/notification/method/base.php]: Is there any valid use case when you do not overwrite these functions in a notification method object?

**Comment by CHItA** [comment]
Overall :+1:

File: phpBB/phpbb/notification/method/base.php (hunk 5 above)
Comment (Nicofuma) [inline phpBB/phpbb/notification/method/base.php]: email, jabber... any method which doesn't store the notification (and so can't interact with it)

File: phpBB/phpbb/notification/type/base.php (hunk 3 above)
Comment (Nicofuma) [inline phpBB/phpbb/notification/type/base.php]: done

**Pull request merged**